modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -110,6 +121,646 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        RexLiteral val = getOperand(pred, RexLiteral.class);
+
+        if (val == null)
+            return guessSelectivity(pred);
+
+        BigDecimal comparableVal = toComparableValue(val);
+
+        if (comparableVal == null)
+            return guessSelectivity(pred);
+
+        return estimateEqualsSelectivity(colStat, comparableVal);
+    }
+
+    /**
+     * Compute selectivity for "is not null" condition.
+     *
+     * @param rel RelNode.
+     * @param mq  RelMetadataQuery.
+     * @param tbl IgniteTable.
+     * @param pred RexNode.
+     * @return Selectivity estimation.
+     */
+    private double estimateIsNotNullSelectivity(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateNotNullSelectivity(colStat);
+    }
+
+    /**
+     * Estimate selectivity for comparison predicate.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable.
+     * @param pred RexNode.
+     * @return Selectivity.
+     */
+    private double estimateComparisonSelectivity(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateRangeSelectivity(colStat, pred);
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred Predicate to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStat(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        SqlKind predKind = pred.getKind();
+
+        if (predKind != SqlKind.IS_NULL && predKind != SqlKind.IS_NOT_NULL && predKind != SqlKind.LOCAL_REF &&
+            predKind != SqlKind.NOT && !predKind.belongsTo(SqlKind.COMPARISON))
+            return null;
+
+        RexLocalRef operand = getOperand(pred, RexLocalRef.class);
+
+        if (operand == null)
+            return null;
+
+        return getColStatBySlot(rel, mq, tbl, operand);
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred RelSlot to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStatBySlot(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexSlot pred) {
+        Set<RelColumnOrigin> origins = mq.getColumnOrigins(rel, pred.getIndex());
```

Comment:
according to `org.apache.calcite.rel.metadata.RelMetadataQuery#getColumnOrigins`'s javadoc, an index is a `0-based ordinal for output column of interest`, thus you can't pass the index of LocalRef here